client_test.go

```diff
@@ -70,55 +70,80 @@ func TestFilenameResolution(t *testing.T) {
+		{1048576, true, false, sha256.New(), "fbbab289f7f94b25736c58be46a994c441fd02552cc6022352e3d86d2fab7c82", false},
+		{128, true, false, sha512.New(), "1dffd5e3adb71d45d2245939665521ae001a317a03720a45732ba1900ca3b8351fc5c9b4ca513eba6f80bc7b1d1fdad4abd13491cb824d61b08d8c0e1561b3f7", true},
+		{128, true, false, sha512.New(), "1dffd5e3adb71d45d2245939665521ae001a317a03720a45732ba1900ca3b8351fc5c9b4ca513eba6f80bc7b1d1fdad4abd13491cb824d61b08d8c0e1561b3f8", false},
+		{1024, true, false, sha512.New(), "37f652be867f28ed033269cbba201af2112c2b3fd334a89fd2f757938ddee815787cc61d6e24a8a33340d0f7e86ffc058816b88530766ba6e231620a130b566c", true},
+		{1024, true, false, sha512.New(), "37f652bf867f28ed033269cbba201af2112c2b3fd334a89fd2f757938ddee815787cc61d6e24a8a33340d0f7e86ffc058816b88530766ba6e231620a130b566d", false},
+		{1048576, true, false, sha512.New(), "ac1d097b4ea6f6ad7ba640275b9ac290e4828cd760a0ebf76d555463a4f505f95df4f611629539a2dd1848e7c1304633baa1826462b3c87521c0c6e3469b67af", true},
+		{1048576, true, false, sha512.New(), "ac1d097c4ea6f6ad7ba640275b9ac290e4828cd760a0ebf76d555463a4f505f95df4f611629539a2dd1848e7c1304633baa1826462b3c87521c0c6e3469b67af", false},
+		// Some cases  false,without Content-Length
+		{128, false, false, md5.New(), "37eff01866ba3f538421b30b7cbefcac", true},
+		{128, false, true, md5.New(), "37eff01866ba3f538421b30b7cbefcac", true}, // resume version
+		{128, false, false, md5.New(), "37eff01866ba3f538421b30b7cbefcad", false},
+		{1024, false, false, md5.New(), "b2ea9f7fcea831a4a63b213f41a8855b", true},
+		{1024, false, false, md5.New(), "b2ea9f7fcea831a4a63b213f41a8855c", false},
+		{1024, false, true, md5.New(), "b2ea9f7fcea831a4a63b213f41a8855c", false}, // resume version
 	}
 
-	for _, test := range tests {
+	for i, test := range tests {
 		comparison := "Match"
 		if !test.match {
 			comparison = "Mismatch"
 		}
+		if test.resume {
+			comparison = comparison + "Resume"
+		}
 
-		t.Run(fmt.Sprintf("With%s%s", comparison, test.sum[:8]), func(t *testing.T) {
-			filename := fmt.Sprintf(".testChecksum-%s-%s", comparison, test.sum[:8])
+		t.Run(fmt.Sprintf("%d-With%s-%s", i, comparison, test.sum[:8]), func(t *testing.T) {
+			filename := fmt.Sprintf(".testChecksum-%d-%s-%s", i, comparison, test.sum[:8])
 			defer os.Remove(filename)
 
 			b, _ := hex.DecodeString(test.sum)
-			req, _ := NewRequest(filename, ts.URL+fmt.Sprintf("?size=%d", test.size))
+
+			var req *Request
+			var resp *Response
+
+			if test.resume {
+				// Do first "uncomplete" request (smaller by 1 byte)
+				req, _ = NewRequest(filename, ts.URL+fmt.Sprintf("?size=%d&sendlength=%v", test.size-1, test.sendlength))
+				req.deleteOnError = false
+				req.SetChecksum(test.hash, b, true)
+				resp = DefaultClient.Do(req)
+			}
+
+			// Normal request
+			req, _ = NewRequest(filename, ts.URL+fmt.Sprintf("?size=%d&sendlength=%v", test.size, test.sendlength))
 			req.SetChecksum(test.hash, b, true)
```

Comment:
Reuse of `test-hash` here creates a data race between the two `DefaultClient.Do` calls.